SNAKE-20 Основной движок игры внутри Canvas

Totentu opened the pull request.

Totentu committed:
Добавил столкновение с краями, потерю здоровья, визуализацию ударов и переход в режим призрака при падении здоровья до 0. Остановку игры, когда все стали призраками.

Totentu committed:
Добавил столкновение с краями, потерю здоровья, визуализацию ударов и переход в режим призрака при падении здоровья до 0. Остановку игры, когда все стали призраками.

Totentu committed:
Добавил столкновение с краями, потерю здоровья, визуализацию ударов и переход в режим призрака при падении здоровья до 0. Остановку игры, когда все стали призраками.

Totentu committed:
Добавил столкновение с краями, потерю здоровья, визуализацию ударов и переход в режим призрака при падении здоровья до 0. Остановку игры, когда все стали призраками.

Totentu committed:
Ядро встроил в рабочую оболочку. Для работы с картинками добавил пакеты
file-loader и image-webpack-loader.

Totentu committed:
Ядро встроил в рабочую оболочку. Для работы с картинками добавил пакеты
file-loader и image-webpack-loader.

ndronov commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "scripts": {
     "dev": "webpack serve --mode development --open --hot ",
-    "build": "webpack --mode production",
```
Почему ты поменял production на development ?
Предполагается что `build` собирает сборку для прода.

Totentu replied:
Мне что-то нужно было проконтролировать внутри сборки, просто забыл вернуть обратно. 

Исправил.

ndronov commented on package.json:
```diff
@@ -36,7 +36,9 @@
     "eslint-plugin-react-hooks": "^4.2.0",
+    "file-loader": "^6.2.0",
     "html-webpack-plugin": "5.5.0",
+    "image-webpack-loader": "^8.0.1",
```
Я настроил сборку картинок - webpack умеет это делать без дополнительных плагинов - он просто копирует их в папку `dist/assets`.

Предлагаю убрать плагины.

ndronov commented on src/components/NavMenu/NavMenu.css:
```diff
@@ -1,4 +1,7 @@
 .nav-menu {
+  width: 100%;
```
я подправил NavMenu, можно не делать эту правку

Totentu committed:
Вернул production для команды build.

ndronov commented on src/components/NavMenu/NavMenu.tsx:
```diff
@@ -15,15 +15,13 @@ function navLinkClass(params: NavLinkClassParams): string {
-          </li>
-        ))}
-      </ul>
+      {routes.map((route) => (
```
это тоже можно отменить

ndronov commented on src/core/classes.ts:
```diff
@@ -0,0 +1,85 @@
+import config from './constants';
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import appleSource from '../assets/apple2.png';
```
Я содал декларации типов для изображений, можно убрать `ts-ignore`

ndronov commented on src/core/classes.ts:
```diff
@@ -0,0 +1,85 @@
+};
+
+class Snake {
+  public Elements: Circle[] = [];
```
поля именуются в camelCase

ndronov commented on src/core/classes.ts:
```diff
@@ -0,0 +1,85 @@
+    // Элементы туловища на 2 пикселя меньше головы, а хвост еще на 2.
+    const dx = vx === 0 ? 0 : (Math.abs(vx) / vx) * config.fieldStep;
+    const dy = vy === 0 ? 0 : (Math.abs(vy) / vy) * config.fieldStep;
+    for (let i = 1; i < config.defaultSnakeLength; i++) {
```
перед циклом для улучшения читабельности лучше оставлять пустую строку

ndronov commented on src/core/constants.ts:
```diff
@@ -0,0 +1,69 @@
+  fieldHeight: 490,
+  fieldLeft: 200,
+  fieldTop: 100,
+  LBorder: 0,
```
лучше с маленькой буквы: `lBorder`

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+const indexOfSnakeUnderControl = 0;
+
+// Функция окончания игры
+function GameOver(): void {
```
функции тоже в camelCase: `gameOver()`

PascalCase только для типов, классов, интерфейсов

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+}
+
+// Функция отрисовки правой панели
+function DrawRightPanel(): void {
```
тут тоже

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+  ctx.closePath();
+
+  ctx.textBaseline = 'top';
+  let textTop = config.rightPanelTop + 10;
```
тут лучше оставить пустую строку

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+
+  ctx.textBaseline = 'top';
+  let textTop = config.rightPanelTop + 10;
+  Snakes.map(function (item) {
```
`map()` используется для создания нового массива, если нужно просто перебрать все элементы, используется `forEach()`

А для колбэка удобнее использовать стрелочную функцию:
```js
array.forEach((item, index) => {
  // do something
})
```

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+let mainTimerStart: number;
+let ctx: CanvasRenderingContext2D;
+const gameStatus = { mode: 'Play' };
+const Snakes: Snake[] = [];
```
экземпляры классов с маленькой буквы:
```js
const snakes: Snake[] = [];
```

ndronov commented on webpack.config.js:
```diff
@@ -19,8 +19,21 @@ module.exports = {
+              disable: true // webpack@2.x and newer
+            }
+          }
+        ]
```
это тоже можно убрать

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+    // Скорость головы растет с учетом оставшегося времени
+    const now = window.performance.now();
+    const pastSeconds = Math.floor((now - mainTimerStart) / 1000);
+    if (head.vx !== 0) {
```
перед и поле `if` лучше оставлять пустую строку для читабельнсти

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+  last = now;
+}
+
+function OKD(e: KeyboardEvent): void {
```
лучше придумать название, корое помогает понять, что делает функция

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+}
+
+// Функция отрисовки монет
+function DrawCoins(): void {
```
с маленькой буквы

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+
+// Функция отрисовки монет
+function DrawCoins(): void {
+  Coins.map(function (el) {
```
лучше forEach вместо map

ndronov commented on src/core/core.ts:
```diff
@@ -0,0 +1,470 @@
+
+document.addEventListener('keydown', OKD);
+
+export { StartGame };
```
очень большой файл получился, я бы разбил его на несколько

Totentu committed:
Мелкие правки по коду.

Totentu committed:
Мелкие правки.

Totentu committed:
Исправил код в соответствии с ревью:
- сделал git rebase на основе обновленного main
- убрал плагины, работающие с картинками
- заменил все полные переборы массивов на forEach
- все названия привел к camelCase
- скорректировал отступы перед if
- расшифровал OKD (onKeyDown)
- разделил файл core.ts на два, отдели от него функции рисования в файл draw.ts

Totentu committed:
Исправил код в соответствии с ревью:
- сделал git rebase на основе обновленного main
- убрал плагины, работающие с картинками
- заменил все полные переборы массивов на forEach
- все названия привел к camelCase
- скорректировал отступы перед if
- расшифровал OKD (onKeyDown)
- разделил файл core.ts на два, отдели от него функции рисования в файл draw.ts

Totentu committed:
Исправил код в соответствии с ревью:
- сделал git rebase на основе обновленного main
- убрал плагины, работающие с картинками
- заменил все полные переборы массивов на forEach
- все названия привел к camelCase
- скорректировал отступы перед if
- расшифровал OKD (onKeyDown)
- разделил файл core.ts на два, отдели от него функции рисования в файл draw.ts

Totentu committed:
Merge remote-tracking branch 'origin/snake-20-main-engine' into snake-20-main-engine

ndronov commented:
да начнётся игра! :snake:
